[1.2] Control system assets through Mix

bennothommo opened the pull request:
This will allow system assets to be installed and updated through Mix, and will also provide us with the ability to be alerted to security advisories for our vendor libraries.

### Libraries swapped:

- Moment.JS replaced with Luxon - Moment is way too bloated for our use at this point where we support major modern browsers in the Backend. In addition, Moment is now in [caretaker mode](https://momentjs.com/docs/#/-project-status/) and its authors themselves recommend switching away from it.

### Upgrades included:

- Bootstrap JS (3.1.1 => 3.4.1)
- jQuery Flot (0.8.2-alpha => 4.2.2)
- Mustache (2.3.2 => 4.2.0)
- Raphael (2.2.7 => 2.3.0)
- Select2 (4.0.3 => 4.0.13)

### Untouched:

- jQuery Clockpicker - appears to have been forked / heavily modified from original source.
- jQuery Mousewheel - based off a modified version from Ayumi that had additional fixes. Could probably remove this entirely TBH - all browsers that we target have proper mousewheel support now.

bennothommo committed:
Delete syntax highlighter used in exception pages

Hasn't been updated for 8 years, and the URL that's listed in the source has been hijacked.

bennothommo committed:
Replace old syntax highlighter with PrismJS

More supported and actively developed. Adds some code highlighting into the error page.

This will be brought in via the System mix config.

bennothommo committed:
Update Select2 through Mix (from 4.0.3 to 4.0.13)

bennothommo committed:
Rebuild assets after NPM update

bennothommo committed:
Update Raphael through Mix (2.2.7 to 2.3.0)

bennothommo committed:
Install Pikaday through Mix (no version change)

bennothommo committed:
Update Mustache through Mix (2.3.2 to 4.2.0)

bennothommo committed:
Swap Moment.js with Luxon 3.0.1

Moment.js is too bloated and was needed when we supported older browsers in the Backend, but since we've phased out support for IE11, Luxon is a superior choice.

bennothommo committed:
Update client lang files to drop Moment config

bennothommo committed:
Update Bootstrap with Mix (3.1.1 to 3.4.1)

LukeTowers commented:
Ooof, not going to accept in 1.2.0, but will review for 1.2.1. Pretty sure this would crash GitHub in the browser if I tried to include these changes in my final review of 1.2 😂 😜

bennothommo committed:
Merge remote-tracking branch 'origin/wip/1.2' into wip/1.2-system-assets

bennothommo committed:
Upgrade jQuery.Flot through Mix (0.8.2-alpha to 0.8.3)

Removes unneeded Flot files that aren't used in Storm build.

bennothommo committed:
Update Storm compiled JS

bennothommo committed:
Use correct Flot repo - update to 4.2.2

bennothommo committed:
Fix Flot build

bennothommo committed:
Update DateTime utility to use Luxon

bennothommo committed:
Tweak relative calendar time formats

bennothommo committed:
Remove test plugin from repo

bennothommo committed:
Swap out original DateTime helper with new Snowboard version

Adds a UI handler for the backend to attach controls

bennothommo committed:
Update build to separate vendor files from Winter backend files

bennothommo committed:
Move Luxon to Backend where it's only used, recompile

bennothommo committed:
Ignore test files for code quality

bennothommo committed:
Harden Snowboard

- The Snowboard and PluginLoader objects are now frozen and cannot be modified.
- Added a Proxy in front of Snowboard to handle plugin loading
- Plugin "Snowboard" instances are blocked from running certain methods
- Update tests to check hardening

bennothommo committed:
Recompile assets

bennothommo committed:
Prevent prototype pollution of PluginLoader and Snowboard

bennothommo committed:
Harden prototypes and abstracts

bennothommo committed:
Recompile

bennothommo committed:
Ignore build files in ESLint

bennothommo committed:
Merge branch 'wip/1.2-harden-snowboard' into wip/1.2-system-assets

bennothommo committed:
Document UI Handler

LukeTowers commented on modules/backend/assets/ui/js/UIHandler.js:
```diff
@@ -1,40 +1,83 @@
-                this.snowboard[plugin](element);
-            });
-        });
+        this.initialisePlugins(document.body);
```
```suggestion
        this.initializePlugins(document.body);
```

bennothommo committed:
Add "ajaxBeforeUpdate" event on elements updated through AJAX

bennothommo committed:
Add disposable elements and observer

bennothommo committed:
Rewrite Balloon Selector into Backend UI

bennothommo committed:
Recompile backend

bennothommo commented:
Closing this PR as I've gone a different route with widget handling and disposing and it's probably too far gone to fix the conflicts here, but will integrate the changes and library updates in another PR.
